File: src/java/tests/detailed/handler/ContextMenuHandler.java
// Copyright (c) 2014 The Chromium Embedded Framework Authors. All rights
// reserved. Use of this source code is governed by a BSD-style license that
// can be found in the LICENSE file.

package tests.detailed.handler;

import java.awt.Frame;

import org.cef.browser.CefBrowser;
import org.cef.callback.CefContextMenuParams;
import org.cef.callback.CefMenuModel;
import org.cef.callback.CefMenuModel.MenuId;
import org.cef.handler.CefContextMenuHandlerAdapter;

import tests.detailed.dialog.SearchDialog;
import tests.detailed.dialog.ShowTextDialog;

public class ContextMenuHandler extends CefContextMenuHandlerAdapter {
  private final Frame owner_;

  public ContextMenuHandler(Frame owner) {
    owner_ = owner;
  }

  @Override
  public void onBeforeContextMenu(CefBrowser browser,
                                  CefContextMenuParams params,
                                  CefMenuModel model) {
    model.addItem(CefMenuModel.MenuId.MENU_ID_FIND, "Find...");
    if (params.hasImageContents())
      model.addItem(CefMenuModel.MenuId.MENU_ID_USER_FIRST, "Download Image...");
  }
  @Override
  public boolean onContextMenuCommand(CefBrowser browser,
                                      CefContextMenuParams params,
                                      int commandId,
                                      int eventFlags) {
    switch (commandId) {
      case MenuId.MENU_ID_VIEW_SOURCE:
        ShowTextDialog visitor = new ShowTextDialog(owner_, "Source of \"" + browser.getURL() + "\"");
        browser.getSource(visitor);
        return true;
      case MenuId.MENU_ID_FIND:
        SearchDialog search = new SearchDialog(owner_, browser);
        search.setVisible(true);
        return true;
      case MenuId.MENU_ID_USER_FIRST:
        browser.startDownload(params.getSourceUrl());
        return true;
      default:
        return false;
    }
  }
}

File: native/context_menu_handler.cpp
// Copyright (c) 2014 The Chromium Embedded Framework Authors. All rights
// reserved. Use of this source code is governed by a BSD-style license that
// can be found in the LICENSE file.

#include "context_menu_handler.h"

#include "jni_util.h"

namespace {

// JNI CefContextMenuParams object.
class ScopedJNIContextMenuParams
    : public ScopedJNIObject<CefContextMenuParams> {
 public:
  ScopedJNIContextMenuParams(JNIEnv* env, CefRefPtr<CefContextMenuParams> obj)
      : ScopedJNIObject<CefContextMenuParams>(
            env,
            obj,
            "org/cef/callback/CefContextMenuParams_N",
            "CefContextMenuParams") {}
};

}  // namespace

ContextMenuHandler::ContextMenuHandler(JNIEnv* env, jobject handler)
    : handle_(env, handler) {}

void ContextMenuHandler::OnBeforeContextMenu(
    CefRefPtr<CefBrowser> browser,
    CefRefPtr<CefFrame> frame,
    CefRefPtr<CefContextMenuParams> params,
    CefRefPtr<CefMenuModel> model) {
  ScopedJNIEnv env;
  if (!env)
    return;

  ScopedJNIBrowser jbrowser(env, browser);
  ScopedJNIFrame jframe(env, frame);
  jframe.SetTemporary();
  ScopedJNIContextMenuParams jparams(env, params);
  jparams.SetTemporary();
  ScopedJNIMenuModel jmodel(env, model);
  jmodel.SetTemporary();

  JNI_CALL_VOID_METHOD(env, handle_, "onBeforeContextMenu",
                       "(Lorg/cef/browser/CefBrowser;Lorg/cef/browser/"
                       "CefFrame;Lorg/cef/callback/CefContextMenuParams;"
                       "Lorg/cef/callback/CefMenuModel;)V",
                       jbrowser.get(), jframe.get(), jparams.get(),
                       jmodel.get());
}

bool ContextMenuHandler::OnContextMenuCommand(
    CefRefPtr<CefBrowser> browser,
    CefRefPtr<CefFrame> frame,
    CefRefPtr<CefContextMenuParams> params,
    int command_id,
    EventFlags event_flags) {
  ScopedJNIEnv env;
  if (!env)
    return false;

  ScopedJNIBrowser jbrowser(env, browser);
  ScopedJNIFrame jframe(env, frame);
  jframe.SetTemporary();
  ScopedJNIContextMenuParams jparams(env, params);
  jparams.SetTemporary();
  jboolean result = JNI_FALSE;

  JNI_CALL_METHOD(env, handle_, "onContextMenuCommand",
                  "(Lorg/cef/browser/CefBrowser;Lorg/cef/browser/CefFrame;Lorg/"
                  "cef/callback/"
                  "CefContextMenuParams;II)Z",
                  Boolean, result, jbrowser.get(), jframe.get(), jparams.get(),
                  (jint)command_id, (jint)event_flags);

  return (result != JNI_FALSE);
}

void ContextMenuHandler::OnContextMenuDismissed(CefRefPtr<CefBrowser> browser,
                                                CefRefPtr<CefFrame> frame) {
  ScopedJNIEnv env;
  if (!env)
    return;

  ScopedJNIBrowser jbrowser(env, browser);
  ScopedJNIFrame jframe(env, frame);
  jframe.SetTemporary();

  JNI_CALL_VOID_METHOD(
      env, handle_, "onContextMenuDismissed",
      "(Lorg/cef/browser/CefBrowser;Lorg/cef/browser/CefFrame;)V",
      jbrowser.get(), jframe.get());
}
